File: PolymorphicEquality/PolymorphicEquality.Models/Address.cs
namespace PolymorphicEquality.Models
{
    public record CountryCode(string Code)
    {
        public override string ToString() => Code;
    }

    public record CityName(string City)
    {
        public static CityName None() => new (string.Empty);
        public override string ToString() => City;
    }

    public record Postcode(string Code)
    {
        public static Postcode None() => new (string.Empty);
        public override string ToString() => Code;
    }

    public record Address(
        CountryCode Country,
        CityName City,
        Postcode Postcode)
    {
        public override string ToString() =>
            $"[Country: '{Country}', City: '{City}', Postcode: '{Postcode}']";
    }
}

File: PolymorphicEquality/PolymorphicEquality.Models.FSharp/Address.fs
module PolymorphicEquality.Models.FSharp.Address

type CountryCode = private CountryCode of string
module CountryCode =
    let create code = CountryCode code
    let value (CountryCode code) = code

type CityName = private CityName of string
module CityName =
    let create city = CityName city
    let value (CityName city) = city

type Postcode = private Postcode of string
module Postcode =
    let create postcode = Postcode postcode
    let value (Postcode postcode) = postcode

type Address = {
    Country: CountryCode
    City: CityName
    Postcode: Postcode
}
